pf/internal/pfutils/proposed_new.go

```diff
@@ -16,217 +16,94 @@ package pfutils
+	if err != nil {
+		return tftypes.Value{}, err
 	}
-	objectType := schema.Type(ctx).(tftypes.Object)
-	if priorState.IsNull() {
-		priorState = newObjectWithDefaults(objectType, func(t tftypes.Type) tftypes.Value {
-			return tftypes.NewValue(t, nil)
-		})
+	conv := &conversion{
+		objectType:    objectType,
+		objectTypeCty: objectTypeCty,
 	}
-
-	joinOpts := JoinOptions{
-		SetElementEqual: NonComputedEq(schema),
-	}
-
-	joinOpts.Reconcile = func(diff Diff) (*tftypes.Value, error) {
-		priorStateValue := diff.Value1
-		configValue := diff.Value2
-		pathType, err := getNearestEnclosingPathType(schema, diff.Path)
-		if err != nil {
-			return nil, fmt.Errorf("%s: %w", diff.Path, err)
-		}
-		switch pathType {
-		case pathToReadOnlyAttribute:
-			return priorStateValue, nil
-		case pathToNonComputedAttribute, pathToBlock:
-			return configValue, nil
-		case pathToRoot:
-			if configValue != nil && !configValue.IsNull() && configValue.IsKnown() {
-				v, err := rewriteNullComputedAsUnknown(schema, diff.Path, *configValue)
-				return &v, err
-			}
-			return priorStateValue, nil
-		case pathToComputedOptionalAttribute:
-			if configValue != nil {
-				if !configValue.IsNull() && configValue.IsKnown() {
-					if priorStateValue != nil && !priorStateValue.IsKnown() {
-						v, err := rewriteNullComputedAsUnknown(schema, diff.Path, *configValue)
-						return &v, err
-					}
-					return configValue, nil
-				}
-				if !configValue.IsKnown() {
-					return configValue, nil
-				}
-			}
-			return priorStateValue, nil
-		default:
-			return nil, fmt.Errorf("joinOpts.Reconcile: impossible pathType case %v on diff=%v", pathType, diff)
-		}
+	priorStateCty, err := conv.ToCtyValue(priorState)
+	if err != nil {
+		return tftypes.Value{}, err
 	}
-
-	joined, err := Join(joinOpts, tftypes.NewAttributePath(), &priorState, &config)
+	configCty, err := conv.ToCtyValue(config)
 	if err != nil {
 		return tftypes.Value{}, err
 	}
-	if joined == nil {
-		return tftypes.Value{}, fmt.Errorf("unexpected nil when computing ProposedNewState")
+	block, err := convertBlock(extractRawSchema(schema))
+	if err != nil {
+		return tftypes.Value{}, err
 	}
-	return *joined, nil
+	proposedNewCty := objchange.ProposedNew(block, priorStateCty, configCty)
+	return conv.FromCtyValue(proposedNewCty)
 }
 
-func rewriteNullComputedAsUnknown(schema tftypes.AttributePathStepper,
-	offset *tftypes.AttributePath, val tftypes.Value) (tftypes.Value, error) {
-	return tftypes.Transform(val, func(p *tftypes.AttributePath, v tftypes.Value) (tftypes.Value, error) {
-		pt, err := getNearestEnclosingPathType(schema, joinPaths(offset, p))
-		if err != nil {
-			return v, err
-		}
-		switch pt {
-		case pathToComputedOptionalAttribute, pathToReadOnlyAttribute:
-			if v.IsNull() {
-				return tftypes.NewValue(v.Type(), tftypes.UnknownValue), nil
-			}
-		}
-		return v, nil
-	})
+func extractRawSchema(schema runtypes.Schema) *tfprotov6.Schema {
+	panic("TODO")
```

Comment:
There is this one remaining hole to fill before tests can succeed. It looks like in https://github.com/pulumi/pulumi-terraform-bridge/blob/master/pf/internal/schemashim/provider.go#L47 we are already having to call server6.GetProviderSchema but discarding the result. We could save the result so we can look up resource schemas by resource ID.